modules/dicehub/extension/action.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// This program is free software: you can use, redistribute, and/or modify
+// it under the terms of the GNU Affero General Public License, version 3
+// or later ("AGPL"), as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package extension
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+
+	"github.com/erda-project/erda-proto-go/core/dicehub/extension/pb"
+	"github.com/erda-project/erda/apistructs"
+)
+
+type Repo struct {
+	// addr workPath
+	addr string
+	// versions ExtensionVersion dir path
+	versions []string
+}
+
+func (s *extensionService) InitExtension(addr string) error {
+	logrus.Infoln("start init extension")
+	// get all extensionVersion in repo
+	repo, err := LoadExtensions(addr)
+	if err != nil {
+		return err
+	}
+
+	// get all extensions existed
+	allActionVersions, err := s.db.QueryAllExtensions()
+	if err != nil {
+		if !gorm.IsRecordNotFoundError(err) {
+			return err
+		}
+	}
+
+	// extensionVersionMap key:ExtensionName value:ExtensionVersion
+	// one extension can have version more then 1
+	extensionVersionMap := make(map[string][]string)
+
+	// extensionTypeMap key:ExtensionName value:ExtensionType
+	extensionTypeMap := make(map[string][]string)
+
+	// add all actionVersions to map
+	for _, v := range allActionVersions {
+		version := extensionVersionMap[v.Name]
+		version = append(version, v.Version)
+		extensionVersionMap[v.Name] = version
+
+		specData := apistructs.Spec{}
+		err = yaml.Unmarshal([]byte(v.Spec), &specData)
+		if err != nil {
+			return err
+		}
+
+		extensionType := extensionTypeMap[v.Name]
+		extensionType = append(extensionType, specData.Type)
+		extensionTypeMap[v.Name] = extensionType
+
+	}
+
+	// push all actionVersions
+	for _, v := range repo.versions {
+		name, version, err := s.RunExtensionsPush(v, extensionVersionMap, extensionTypeMap)
+		if err == nil {
+			logrus.Infoln("extension create success, name: ", name, ", version: ", version)
+		}
+	}
+	return nil
+}
+
+// RunExtensionsPush push extensions
+func (s *extensionService) RunExtensionsPush(dir string, extensionVersionMap, extensionTypeMap map[string][]string) (string, string, error) {
+	workDir := dir
+
+	// read spec.yml
+	specBytes, err := ioutil.ReadFile(path.Join(workDir, "spec.yml"))
+	if err != nil {
+		return "", "", err
+	}
+	specData := apistructs.Spec{}
+	err = yaml.Unmarshal(specBytes, &specData)
+	if err != nil {
+		return "", "", errors.Wrap(err, "failed to parse spec")
+	}
+
+	// if extension is existed, return
```

Comment:
maybe version need be updated

Reply:
Only Create new extension。
if you want to update extension, please use CLI push